File: Sources/YTools/Services/SnippetManager.swift
import AppKit
import Combine
import Foundation
import YToolsCore
import YToolsModuleKit

@MainActor
final class SnippetManager: ObservableObject, SnippetSaving {
    @Published private(set) var items: [SnippetItem]
    @Published private(set) var storageError: String?
    @Published private(set) var isLoaded = false
    @Published private(set) var isSaving = false
    private let writer: OrderedSecureStoreWriter
    private let saveDebouncer = DebouncedAction()
    private var pendingMutations: [PendingMutation] = []
    private var initializationTask: Task<Void, Never>?
    private var revision: UInt64 = 0
    private var queuedRevision: UInt64 = 0
    private var latestSaveTask: Task<Bool, Never>?

    init() {
        self.writer = OrderedSecureStoreWriter { SecureCodableStore(name: "snippets") }
        items = []
        storageError = nil
        startLoading()
    }

    init(storeFactory: @escaping @Sendable () -> SecureCodableStore) {
        self.writer = OrderedSecureStoreWriter(storeFactory: storeFactory)
        items = []
        storageError = nil
        startLoading()
    }

    var storageStatus: String {
        if !isLoaded { return "正在读取加密片段…" }
        if isSaving { return "正在加密保存…" }
        return storageError ?? "已安全保存"
    }

    private func startLoading() {
        initializationTask = Task { [weak self, writer] in
            let result = await writer.load([SnippetItem].self)
            guard let self else { return }
            applyLoad(result)
        }
    }

    private func applyLoad(_ result: BackgroundStoreLoad<[SnippetItem]>) {
        guard !isLoaded else { return }
        var loaded: [SnippetItem]
        switch result {
        case .missing:
            loaded = []
            storageError = nil
        case let .loaded(value):
            loaded = value
            storageError = nil
        case let .unavailable(message), let .corrupted(message):
            loaded = []
            storageError = message
        }
        pendingMutations.forEach { $0.apply(to: &loaded) }
        pendingMutations.removeAll()
        items = loaded
        isLoaded = true
        if revision > 0 { _ = queueCurrentSnapshot(failureMessage: "无法写入加密文本片段；现有文件未被覆盖。") }
    }

    func searchModule(clipboardText: String, now: Date = Date()) -> SnippetSearchModule {
        SnippetSearchModule(items: items, clipboardText: clipboardText, now: now)
    }

    var saveError: String? { storageError }

    @discardableResult
    func save(text: String, title: String? = nil, keyword: String = "", collection: String = "默认") -> Bool {
        beginSave(text: text, title: title, keyword: keyword, collection: collection) != nil
    }

    func savePersisted(text: String, title: String? = nil, keyword: String = "", collection: String = "默认") async -> Bool {
        guard let targetRevision = beginSave(text: text, title: title, keyword: keyword, collection: collection) else { return false }
        await initializationTask?.value
        if queuedRevision < targetRevision {
            return await queueCurrentSnapshot(failureMessage: "无法写入加密文本片段；现有文件未被覆盖。").value
        }
        return await waitForQueuedWrites()
    }

    private func beginSave(text: String, title: String?, keyword: String, collection: String) -> UInt64? {
        let trimmed = text.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return nil }
        let inferredTitle = title?.trimmingCharacters(in: .whitespacesAndNewlines)
        let finalTitle = inferredTitle.flatMap { $0.isEmpty ? nil : $0 } ?? preview(trimmed, limit: 36)
        let now = Date()
        let item = SnippetItem(
            id: UUID(),
            title: finalTitle,
            keyword: keyword,
            content: text,
            collection: collection,
            createdAt: now,
            updatedAt: now
        )
        apply(.insert(item))
        if isLoaded { _ = queueCurrentSnapshot(failureMessage: "无法写入加密文本片段；现有文件未被覆盖。") }
        return revision
    }

    func delete(_ item: SnippetItem) {
        apply(.delete(item.id))
        if isLoaded { _ = queueCurrentSnapshot(failureMessage: "无法保存删除操作。") }
    }

    func update(
        id: UUID,
        title: String? = nil,
        keyword: String? = nil,
        content: String? = nil,
        collection: String? = nil
    ) {
        guard items.contains(where: { $0.id == id }) else { return }
        apply(.update(id, title, keyword, content, collection, Date()))
        saveDebouncer.schedule(after: .milliseconds(350)) { [weak self] in
            guard let self, self.isLoaded else { return }
            _ = self.queueCurrentSnapshot(failureMessage: "无法保存文本片段修改。")
        }
    }

    func flushPendingChanges() async {
        saveDebouncer.cancel()
        await initializationTask?.value
        if queuedRevision < revision {
            _ = await queueCurrentSnapshot(failureMessage: "无法保存文本片段修改。").value
        }
        _ = await waitForQueuedWrites()
    }

    func waitUntilLoaded() async { await initializationTask?.value }

    private func apply(_ mutation: PendingMutation) {
        saveDebouncer.cancel()
        mutation.apply(to: &items)
        if !isLoaded { pendingMutations.append(mutation) }
        revision &+= 1
    }

    private func queueCurrentSnapshot(failureMessage: String) -> Task<Bool, Never> {
        let snapshot = items
        let saveRevision = revision
        queuedRevision = max(queuedRevision, saveRevision)
        isSaving = true
        let previous = latestSaveTask
        let task = Task { [weak self, writer] in
            _ = await previous?.value
            let success = await writer.save(snapshot)
            guard let self else { return success }
            if saveRevision == revision {
                isSaving = false
                storageError = success ? nil : failureMessage
            }
            return success
        }
        latestSaveTask = task
        return task
    }

    private func waitForQueuedWrites() async -> Bool {
        // Submitting a sentinel save is unnecessary: every returned save task
        // completes only after all earlier actor operations.
        if queuedRevision < revision { return false }
        guard let latestSaveTask else { return queuedRevision == revision && storageError == nil }
        return await latestSaveTask.value
    }

    private enum PendingMutation {
        case insert(SnippetItem)
        case delete(UUID)
        case update(UUID, String?, String?, String?, String?, Date)

        func apply(to items: inout [SnippetItem]) {
            switch self {
            case let .insert(item): items.insert(item, at: 0)
            case let .delete(id): items.removeAll { $0.id == id }
            case let .update(id, title, keyword, content, collection, date):
                guard let index = items.firstIndex(where: { $0.id == id }) else { return }
                if let title { items[index].title = title }
                if let keyword { items[index].keyword = keyword }
                if let content { items[index].content = content }
                if let collection { items[index].collection = collection }
                items[index].updatedAt = date
            }
        }
    }

    private func preview(_ text: String, limit: Int = 72) -> String {
        let singleLine = text.replacingOccurrences(of: "\n", with: " ")
        return singleLine.count <= limit ? singleLine : String(singleLine.prefix(limit)) + "…"
    }
}

struct SnippetSearchModule: YToolsModule {
    let descriptor = ModuleDescriptor(id: "snippets", name: "文本片段")
    let items: [SnippetItem]
    let clipboardText: String
    let now: Date

    static func accepts(_ query: String) -> Bool {
        searchTerm(query) != nil
    }

    func search(_ request: ModuleSearchRequest) async throws -> [LauncherResult] {
        guard let term = searchTerm(request.query) else { return [] }
        return items.filter { item in
            term.isEmpty
                || item.title.localizedCaseInsensitiveContains(term)
                || item.keyword.localizedCaseInsensitiveContains(term)
                || item.content.localizedCaseInsensitiveContains(term)
                || item.collection.localizedCaseInsensitiveContains(term)
        }
        .prefix(30)
        .map { item in
            let expanded = expand(item.content)
            return LauncherResult(
                id: "snippet:\(item.id.uuidString)",
                moduleID: descriptor.id,
                title: item.title,
                subtitle: [item.collection, item.keyword, preview(expanded)]
                    .filter { !$0.isEmpty }
                    .joined(separator: " · "),
                icon: .system("text.quote"),
                score: 820,
                action: .copy(expanded)
            )
        }
    }

    private static func searchTerm(_ query: String) -> String? {
        let trimmed = query.trimmingCharacters(in: .whitespacesAndNewlines)
        for prefix in ["snip", "snippet", "片段"] {
            if trimmed.lowercased() == prefix { return "" }
            if trimmed.lowercased().hasPrefix(prefix + " ") {
                return String(trimmed.dropFirst(prefix.count + 1))
            }
        }
        return nil
    }

    private func searchTerm(_ query: String) -> String? { Self.searchTerm(query) }

    private func expand(_ content: String) -> String {
        let formatter = DateFormatter()
        formatter.locale = .current
        formatter.dateStyle = .medium
        let date = formatter.string(from: now)
        formatter.dateStyle = .none
        formatter.timeStyle = .medium
        let time = formatter.string(from: now)
        return content
            .replacingOccurrences(of: "{date}", with: date)
            .replacingOccurrences(of: "{time}", with: time)
            .replacingOccurrences(of: "{clipboard}", with: clipboardText)
            .replacingOccurrences(of: "{cursor}", with: "")
    }

    private func preview(_ text: String, limit: Int = 72) -> String {
        let singleLine = text.replacingOccurrences(of: "\n", with: " ")
        return singleLine.count <= limit ? singleLine : String(singleLine.prefix(limit)) + "…"
    }
}

File: src/YTools.Windows/Services/SnippetManager.cs
using System.Globalization;
using System.Windows;
using YTools.Core;
using YTools.Infrastructure;
using YTools.Models;
using YTools.ModuleKit;
using YTools.Services.Storage;

namespace YTools.Services;

public sealed class SnippetManager : ObservableObject
{
    private readonly Func<SecureCodableStore> _storeFactory;
    private readonly OrderedBackgroundWriter _writer = new();
    private readonly DebouncedAction _saveDebouncer = new();
    private readonly List<Action<List<SnippetItem>>> _pendingLoadMutations = [];
    private List<SnippetItem> _items = [];
    private string? _storageError;
    private SecureCodableStore? _store;
    private bool _isLoaded;
    private bool _isSaving;
    private long _revision;
    private long _queuedRevision;
    private readonly Task _initialization;
    private SecureStoreLoadResult<List<SnippetItem>>? _loadedResult;
    private Task<bool>? _latestSaveTask;

    public SnippetManager()
        : this(() => new SecureCodableStore("snippets"))
    {
    }

    internal SnippetManager(Func<SecureCodableStore> storeFactory)
    {
        _storeFactory = storeFactory;
        _initialization = _writer.Enqueue(LoadOnWorker);
    }

    private bool LoadOnWorker()
    {
        SecureStoreLoadResult<List<SnippetItem>> result;
        try
        {
            _store = _storeFactory();
            result = _store.Load<List<SnippetItem>>();
        }
        catch (Exception exception)
        {
            result = SecureStoreLoadResult<List<SnippetItem>>.Unavailable(exception.Message);
        }
        _loadedResult = result;
        PostToOwner(() => ApplyLoad(result));
        return true;
    }

    private void ApplyLoad(SecureStoreLoadResult<List<SnippetItem>> result)
    {
        if (_isLoaded) { return; }
        var loaded = result.Kind == SecureStoreLoadResultKind.Loaded ? result.Value ?? [] : [];
        switch (result.Kind)
        {
            case SecureStoreLoadResultKind.Missing:
                StorageError = null;
                break;
            case SecureStoreLoadResultKind.Loaded:
                StorageError = null;
                break;
            default:
                StorageError = result.Message;
                break;
        }

        foreach (var mutation in _pendingLoadMutations)
        {
            mutation(loaded);
        }

        _pendingLoadMutations.Clear();
        _items = loaded;
        _isLoaded = true;
        RaisePropertyChanged(nameof(Items));
        RaisePropertyChanged(nameof(IsLoaded));
        RaisePropertyChanged(nameof(StorageStatus));
        if (_revision > 0)
        {
            _ = QueueCurrentSnapshot("无法写入加密文本片段；现有文件未被覆盖。");
        }
    }

    public IReadOnlyList<SnippetItem> Items => _items;

    public string? StorageError
    {
        get => _storageError;
        private set
        {
            if (SetField(ref _storageError, value)) { RaisePropertyChanged(nameof(StorageStatus)); }
        }
    }

    public string? SaveError => StorageError;

    public bool IsLoaded => _isLoaded;

    public bool IsSaving
    {
        get => _isSaving;
        private set
        {
            if (SetField(ref _isSaving, value))
            {
                RaisePropertyChanged(nameof(StorageStatus));
            }
        }
    }

    public string StorageStatus => !_isLoaded ? "正在读取加密片段…" : IsSaving ? "正在加密保存…" : StorageError ?? "已安全保存";

    public SnippetSearchModule SearchModule(string clipboardText, DateTimeOffset now)
    {
        return new SnippetSearchModule(_items.ToList(), clipboardText, now);
    }

    public bool Save(string text, string? title = null, string keyword = "", string collection = "默认")
    {
        return BeginSave(text, title, keyword, collection, out _) is not null;
    }

    public Task<bool> SaveAsync(string text, string? title = null, string keyword = "", string collection = "默认")
    {
        return BeginSave(text, title, keyword, collection, out var revision)
            ?? Task.FromResult(false);
    }

    private Task<bool>? BeginSave(
        string text,
        string? title,
        string keyword,
        string collection,
        out long revision)
    {
        revision = 0;
        var trimmed = text.Trim();
        if (string.IsNullOrEmpty(trimmed))
        {
            return null;
        }

        var finalTitle = string.IsNullOrWhiteSpace(title)
            ? Preview(trimmed)
            : title.Trim();
        var now = DateTimeOffset.UtcNow;
        var newItem = new SnippetItem(
                Guid.NewGuid(),
                finalTitle,
                keyword,
                text,
                collection,
                now,
                now);
        ApplyMutation(items => items.Insert(0, newItem));
        revision = _revision;
        return _isLoaded
            ? QueueCurrentSnapshot("无法写入加密文本片段；现有文件未被覆盖。")
            : WaitForRevisionAsync(revision);
    }

    public void Delete(SnippetItem item)
    {
        ApplyMutation(items => items.RemoveAll(existing => existing.Id == item.Id));
        if (_isLoaded) { _ = QueueCurrentSnapshot("无法保存删除操作。"); }
    }

    public void Update(
        Guid id,
        string? title = null,
        string? keyword = null,
        string? content = null,
        string? collection = null)
    {
        if (_items.All(item => item.Id != id))
        {
            return;
        }

        ApplyMutation(items =>
        {
            var index = items.FindIndex(item => item.Id == id);
            if (index < 0) { return; }
            var item = items[index];
            items[index] = item with
            {
                Title = title ?? item.Title,
                Keyword = keyword ?? item.Keyword,
                Content = content ?? item.Content,
                Collection = collection ?? item.Collection,
                UpdatedAt = DateTimeOffset.UtcNow
            };
        });
        _saveDebouncer.Schedule(TimeSpan.FromMilliseconds(350), () =>
        {
            if (_isLoaded) { _ = QueueCurrentSnapshot("无法保存文本片段修改。"); }
        });
    }

    public void FlushPendingChanges()
    {
        _saveDebouncer.Cancel();
        _initialization.GetAwaiter().GetResult();
        if (!_isLoaded && _loadedResult is { } loaded) { ApplyLoad(loaded); }
        if (_queuedRevision < _revision) { _ = QueueCurrentSnapshot("无法保存文本片段修改。"); }
        _writer.DrainAsync().GetAwaiter().GetResult();
    }

    public async Task FlushPendingChangesAsync()
    {
        _saveDebouncer.Cancel();
        await _initialization.ConfigureAwait(false);
        Task<bool>? save = null;
        PostToOwner(() =>
        {
            if (!_isLoaded && _loadedResult is { } loaded) { ApplyLoad(loaded); }
            save = _queuedRevision < _revision
                ? QueueCurrentSnapshot("无法保存文本片段修改。")
                : _latestSaveTask;
        }, synchronous: true);
        if (save is not null) { _ = await save.ConfigureAwait(false); }
        await _writer.DrainAsync().ConfigureAwait(false);
    }

    internal Task WaitUntilLoadedAsync() => _initialization;

    private void ApplyMutation(Action<List<SnippetItem>> mutation)
    {
        _saveDebouncer.Cancel();
        mutation(_items);
        if (!_isLoaded) { _pendingLoadMutations.Add(mutation); }
        _revision += 1;
        RaisePropertyChanged(nameof(Items));
    }

    private Task<bool> QueueCurrentSnapshot(string failureMessage)
    {
        var snapshot = _items.ToList();
        var revision = _revision;
        _queuedRevision = Math.Max(_queuedRevision, revision);
        IsSaving = true;
        var task = _writer.Enqueue(() => _store?.Save(snapshot) == true);
        _latestSaveTask = task;
        _ = task.ContinueWith(completed => PostToOwner(() =>
        {
            if (revision != _revision) { return; }
            IsSaving = false;
            StorageError = completed.IsCompletedSuccessfully && completed.Result ? null : failureMessage;
            RaisePropertyChanged(nameof(StorageStatus));
        }), TaskScheduler.Default);
        return task;
    }

    private async Task<bool> WaitForRevisionAsync(long revision)
    {
        await _initialization.ConfigureAwait(false);
        Task<bool>? save = null;
        PostToOwner(() =>
        {
            if (!_isLoaded && _loadedResult is { } loaded) { ApplyLoad(loaded); }
            save = _queuedRevision >= revision
                ? _latestSaveTask
                : QueueCurrentSnapshot("无法写入加密文本片段；现有文件未被覆盖。");
        }, synchronous: true);
        return save is not null && await save.ConfigureAwait(false);
    }

    private static void PostToOwner(Action action, bool synchronous = false)
    {
        var dispatcher = Application.Current?.Dispatcher;
        if (dispatcher is not null && !dispatcher.CheckAccess())
        {
            if (synchronous) { dispatcher.Invoke(action); }
            else { _ = dispatcher.BeginInvoke(action); }
        }
        else { action(); }
    }

    private static string Preview(string text, int limit = 72)
    {
        var singleLine = text.Replace("\n", " ");
        return singleLine.Length <= limit ? singleLine : singleLine[..limit] + "…";
    }
}

public sealed class SnippetSearchModule : IYToolsModule
{
    private readonly IReadOnlyList<SnippetItem> _items;
    private readonly string _clipboardText;
    private readonly DateTimeOffset _now;

    public SnippetSearchModule(IReadOnlyList<SnippetItem> items, string clipboardText, DateTimeOffset now)
    {
        _items = items;
        _clipboardText = clipboardText;
        _now = now;
    }

    public ModuleDescriptor Descriptor { get; } = new("snippets", "文本片段");

    public static bool Accepts(string query)
    {
        return SearchTerm(query) is not null;
    }

    public Task<IReadOnlyList<LauncherResult>> SearchAsync(ModuleSearchRequest request)
    {
        var term = SearchTerm(request.Query);
        if (term is null)
        {
            return Task.FromResult<IReadOnlyList<LauncherResult>>([]);
        }

        var results = _items
            .Where(item =>
                string.IsNullOrEmpty(term)
                || item.Title.Contains(term, StringComparison.OrdinalIgnoreCase)
                || item.Keyword.Contains(term, StringComparison.OrdinalIgnoreCase)
                || item.Content.Contains(term, StringComparison.OrdinalIgnoreCase)
                || item.Collection.Contains(term, StringComparison.OrdinalIgnoreCase))
            .Take(30)
            .Select(item =>
            {
                var expanded = Expand(item.Content);
                return new LauncherResult(
                    $"snippet:{item.Id:N}",
                    Descriptor.Id,
                    item.Title,
                    string.Join(" · ", new[] { item.Collection, item.Keyword, Preview(expanded) }
                        .Where(part => !string.IsNullOrEmpty(part))),
                    new ResultIcon.System("text.quote"),
                    820,
                    new ResultAction.Copy(expanded));
            }).Cast<LauncherResult>().ToList();

        return Task.FromResult<IReadOnlyList<LauncherResult>>(results);
    }

    private static string? SearchTerm(string query)
    {
        var trimmed = query.Trim();
        foreach (var prefix in new[] { "snip", "snippet", "片段" })
        {
            if (string.Equals(trimmed, prefix, StringComparison.OrdinalIgnoreCase))
            {
                return "";
            }

            if (trimmed.StartsWith(prefix + " ", StringComparison.OrdinalIgnoreCase))
            {
                return trimmed[(prefix.Length + 1)..];
            }
        }

        return null;
    }

    private string Expand(string content)
    {
        var date = _now.ToLocalTime().ToString("yyyy/M/d", CultureInfo.InvariantCulture);
        var time = _now.ToLocalTime().ToString("HH:mm:ss", CultureInfo.InvariantCulture);
        return content
            .Replace("{date}", date)
            .Replace("{time}", time)
            .Replace("{clipboard}", _clipboardText)
            .Replace("{cursor}", "");
    }

    private static string Preview(string text, int limit = 72)
    {
        var singleLine = text.Replace("\n", " ");
        return singleLine.Length <= limit ? singleLine : singleLine[..limit] + "…";
    }
}
